### DISCUSSION TITLE
Example for assigning new values to  FITS Headers?

### DISCUSSION THREAD

**Discussion by JuergenTerpe** [brampf/fitscore#19] [Q&A]
Hi, 

I'm trying to assign new values with a custom keyword to the FITS headers and it seems I'm doing it wrong, obviously (despite it seems to be intuitive). Do you have a working example how to do so? Here is my code, perhaps you can see immediately what I'm doing wrong? 

           if let fits = try? FitsFile.read(contentsOf: image.url) {
			let header = fits.prime.headerUnit
			
			if settings.hfr, let hfr = image.hfr {
				header["HFR"] = hfr
			}
			
			if settings.fwhm, let fwhm = image.fwhm {
				header["FWHM"] = fwhm
			}
			
			if settings.eccentricity, let ecc = image.eccentricity {
				header["ECCEN"] = ecc
			}
			
			if settings.mean, let mean = image.mean {
				header["MEAN"] = mean
			}
			
			if settings.median, let median = image.median {
				header["MEDIAN"] = median
			}
			
			if settings.aduMin, let aduMin = image.aduMin {
				header["ADUMIN"] = aduMin
			}
			
			if settings.aduMax, let aduMax = image.aduMax {
				header["ADUMAX"] = aduMax
			}
			
			if settings.numStars, let numStars = image.numStars {
				header["NSTARS"] = numStars
			}
			
			if settings.rms, let rms = image.rms {
				header["RMS"] = rms
			}
			
			if settings.rmsRA, let rmsRA = image.rmsRA {
				header["RMSRA"] = rmsRA
			}
			
			if settings.rmsDEC, let rmsDEC = image.rmsDEC {
				header["RMSDEC"] = rmsDEC
			}
			
			if let pathURL = settings.path {
				let url = pathURL.appendingPathComponent(image.FileName)
				fits.write(to: url) { error in
					print(error.localizedDescription)
				} onCompleation: {
					onProgress()
				}
			} else {
				fits.write(to: image.url) { error in
					print(error.localizedDescription)
				} onCompleation: {
					onProgress()
				}
			}
		}

The FITS file will be written successfully, but PixInsight cannot open it - and also not FITScore. PixeInsight gives me this error message: 

*** Error: PCL Legacy FITS Format Support: Unable to open FITS file: /Volumes/ImageStore/Leo Triplett/Temp/LIGHT/2024-04-11_00-33-12_Baader Neodymium_-4.90_180.00s_0035.fits
CFITSIO error message stack: 
01 : Character 9 in this keyword is illegal. Hex Value = FFFFFF80
02 : ÄÚÄÝÇæÊç½éÅ¼õ¿èÀôÊæ»ó¿ð¿ëÁîÇò¼ñÂäÁòÄ
03 : Name of keyword no. 109 contains illegal character(s): ÄÚÄÝÇæÊç½éÅ
04 : Failed to read the required primary array header keywords.
05 : ffopen could not interpret primary array header of file: 
06 : /Volumes/ImageStore/Leo Triplett/Temp/LIGHT/2024-04-11_00-33-12_Baader Neodymium
07 : _-4.90_180.00s_0035.fits
*** Unable to open file: /Volumes/ImageStore/Leo Triplett/Temp/LIGHT/2024-04-11_00-33-12_Baader Neodymium_-4.90_180.00s_0035.fits
<* failed *>
 
 Thanks in advance for your help!

**Comment by JuergenTerpe** [comment]
I also tried this with strings.... Maybe, all this is not the intention how to use it.

**Reply by brampf** [reply]
Hi there!
haven't been working on this code for quite a while, so this might take a while. Can you share the file you are working on before and after you made your changes so I can have a look at it?

**Reply by JuergenTerpe** [reply]
Sure, I can do. 

Some background information: I' writing an analysis tool to evaluate complete imaging sessions captured with NINA. There is an additional plugin available, which writes some important info into a csv file, such as HFR or star eccentricity. I wanted to provide an option to add this information into the FITS header for later use in PixInsight. This free and open source tool will use FitsCore. 
This can be a very helpful tool for astrophotographers worldwide - at least for the Mac users. 

I don't think your library is incorrect, I would think I don't do it right. 

Regards and Thanks for having a look,
Jürgen


[After Writing Header with new Data.fits.zip](https://github.com/brampf/fitscore/files/15026112/After.Writing.Header.with.new.Data.fits.zip)
[Original.fits.zip](https://github.com/brampf/fitscore/files/15026125/Original.fits.zip)
